### PR TITLE
First building blocks for Sharding API: ShardIndex and TransportIdentity

### PR DESCRIPTION
**Pull request by akoshelev**
See #969 for overall approach.

This change introduces two fundamental concepts for horizontal scaling - `ShardIndex` and `Transportidentity`. 

The former will be used the same way as `HelperIdentity` is used nowadays - unique identifier of a peer we are communicating with. Having in mind 200B events, we may need more than 65k shards, thus 4 byte identifier usage. 

`TransportIdentity` is a generalization for peer identifiers used at transport layer.  Two fundamental structs implement it: `HelperIdentity` and `ShardIndex` to allow communication via MPC and shard channels.

`Role` is used at the Gateway/Context layer for communication purposes and `RoleResolvingTransport` is a convenient abstraction, so it also implements `TransportIdentity`.

`Transport` trait became generic over `TransportIdentity` - I tested it in my private branch where I have a complete implementation, so this approach works. 

I also did a bit of a refactoring and made `RoleResolvingTransport` to implement the `Transport` trait because it can now.

### CODE REVIEW

File: ipa-core/src/helpers/gateway/transport.rs [hunk 1]
```diff
@@ -1,64 +1,94 @@
...
+pub struct RoleResolvingTransport {
+    pub(super) roles: RoleAssignment,
+    pub(super) inner: TransportImpl,
+}
+
+impl Stream for RoleRecordsStream {
+    type Item = Vec<u8>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.project().0.poll_next(cx)
```
Comment (benjaminsavage) [inline ipa-core/src/helpers/gateway/transport.rs:39]: What is `project()` and where is it coming from? Is it defined in `Stream`?

File: ipa-core/src/helpers/mod.rs [hunk 2]
```diff
@@ -405,23 +406,25 @@ impl TryFrom<[Role; 3]> for RoleAssignment {
 /// Combination of helper role and step that uniquely identifies a single channel of communication
 /// between two helpers.
 #[derive(Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
-pub struct ChannelId {
-    pub role: Role,
+pub struct ChannelId<I: transport::Identity> {
+    pub peer: I,
```
Comment (benjaminsavage) [inline ipa-core/src/helpers/mod.rs:410]: This name: `peer` confuses me. Why did you change this from `role`? What does it mean here? And can you add a comment?

File: ipa-core/src/sharding.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+pub trait ShardConfiguration {
+    /// Returns the index of the current shard.
+    fn shard_index(&self) -> ShardIndex;
+
+    /// Total number of shards present on this helper. It is expected that all helpers have the
+    /// same number of shards.
+    fn shard_count(&self) -> ShardIndex;
+
+    /// Returns an iterator that yields shard indices for all shards present in the system, except
+    /// this once. Shards are yielded in ascending order.
```
Comment (benjaminsavage) [inline ipa-core/src/sharding.rs:24]: Do you mean "except this one?"

**Review by benjaminsavage — APPROVED** [review]
This looks pretty straightforward to me. Most of the lines are just replacing ChanelId with HelperChanelId.

File: ipa-core/src/helpers/mod.rs (hunk 2 above)
Comment (akoshelev) [reply ipa-core/src/helpers/mod.rs:410]: the reason I had to rename it is that this struct is now used to communicate between shards too and there is no notion of helper role for such communications. `ChannelId<ShardIndex>` would have `role: ShardIndex` field which looks even more confusing, hence a neutral `peer` that can be anything that wants to communicate with this instance

File: ipa-core/src/sharding.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    /// if current shard index is greater or equal to the total number of shards.
+    fn peer_shards(&self) -> impl Iterator<Item = ShardIndex> {
+        let this = self.shard_index();
+        let max = self.shard_count();
+        assert!(
+            this < max,
+            "Current shard index '{this}' >= '{max}' (total number of shards)"
+        );
+
+        max.iter().filter(move |&v| v != this)
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply ipa-core/src/sharding.rs:36]: it moves `this` to the closure, so Rust compiler can see that `filter` closure cannot outlive it. We return an iterator and `filter` closure is part of it, so Rust wants to know that this closure does not borrow from local context
